### PR TITLE
fix(ids): prevent duplicate ids

### PR DESCRIPTION
**Pull request by ItayBenMoshe**

### CODE REVIEW

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 1]
```diff
@@ -1,14 +1,48 @@
...
+
 const useDuplicateContactId = () => {
     
-    const {alertWarning} = useCustomSwal();
+    const { alertWarning } = useCustomSwal();
+    const epidemiologyNumber = useSelector<StoreStateType, number>(state => state.investigation.epidemiologyNumber);
+    const userId = useSelector<StoreStateType, string>(state => state.user.id);
+
+    const checkDuplicateIds = (idsToCheck: string[]) => {
+        const trimmedIds: string[] = idsToCheck.filter((id: string) => id !== null && id !== '');
```
Comment (MayaSasson) [inline client/src/Utils/vendor/useDuplicateContactId.ts:22]: You can jus check !id

Comment (MayaSasson) [reply client/src/Utils/vendor/useDuplicateContactId.ts:22]: In addition you can just filter twice into the same array

File: client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx [hunk 2]
```diff
@@ -56,42 +47,13 @@ const InvestigationForm: React.FC = (): JSX.Element => {
...
+        lastTabDisplayedId
     } = useTabManagement();
 
-    const lastTabDisplayedId = areThereContacts ? LAST_TAB_ID : LAST_TAB_ID - 1;
-
-    const isInvestigationValid = () => {
-        return !(formsValidations.slice(0, lastTabDisplayedId + 1).some((formValidation) => !formValidation));
-    }
-
-    const handleNextPageClick = () => {
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx:71]: Why did you delete it?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 3]
```diff
@@ -50,33 +54,46 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
+                    service: Service.CLIENT,
+                    severity: Severity.LOW,
+                    workflow: 'Saving Contacted People Excel',
+                    step: `trying to save excel with complete status contact`,
+                    investigation: epidemiologyNumber,
+                    user: userId
                 })
+                alertError('לא ניתן לטעון מגע בסטטוס הושלם התחקור, שנו סטטוס זה ונסו שוב');
+            }
+            else {
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:70]: You can combine the else if

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts [hunk 4]
```diff
@@ -68,27 +71,52 @@ const useContactQuestioning = (parameters: useContactQuestioningParameters): use
...
+                        workflow: 'Saving all contacts',
+                        step: 'got respond from the server',
+                        user: userId,
+                        investigation: epidemiologyNumber
+                    });
+                }
+            }).catch(err => {
+                logger.error({
+                    service: Service.CLIENT,
+                    severity: Severity.LOW,
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:112]: severity high

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 5]
```diff
@@ -18,16 +19,18 @@ const fileEndings = [
...
 interface ExcelUploaderProps {
-    contactEvent:number;
+    contactEvent: number;
     onSave: () => void;
+    allInteractions: Interaction[];
 }
-const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
+const ContactUploader = ({contactEvent, onSave, allInteractions} :ExcelUploaderProps) => {
+    const existingContacts = allInteractions.map(interaction => interaction.contacts).flat();
+    const existingContactsIds : string[] = existingContacts.map(contact => contact.idNumber).filter(id => id !== null && id !== '') as string[];
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:28]: check !id

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 6]
```diff
@@ -50,33 +54,46 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
-                        service: Service.CLIENT,
-                        severity: Severity.LOW,
-                        workflow: 'Saving Contacted People Excel',
-                        step: `got error from server: ${error}`,
-                        investigation: epidemiologyNumber,
-                        user: userId
-                    })
-                    alertError('שגיאה בשמירת הנתונים');
+            const excelContactsIds = contacts.map((contact) => contact.identificationNumber)
+                                             .filter(id => id !== null && id !== '');
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:58]: check !id

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 7]
```diff
@@ -50,33 +54,46 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
+            if(contacts.some((contact) => contact.contactStatus == 'הושלם התחקור')) {
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:59]: export to const

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 8]
```diff
@@ -50,33 +54,46 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
+                                    step: 'contacted people excel was saved successfully',
+                                    investigation: epidemiologyNumber,
+                                    user: userId
+                                })
+                                onSave();
+                        })
+                        .catch(error => {
+                            logger.error({
+                                service: Service.CLIENT,
+                                severity: Severity.LOW,
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:87]: change to high

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx [hunk 9]
```diff
@@ -76,7 +79,25 @@ const InteractionEventForm: React.FC<Props> = (
 
   const onSubmit = (data: InteractionEventDialogData) => {
     const interactionDataToSave = convertData(data);
-    saveInteractions(interactionDataToSave);
+    let allContactsIds: IdToCheck[] = [];
+    let newIds: IdToCheck[] = [];
+    interactions.forEach((interaction: InteractionEventDialogData) => {
```
Comment (NaamanTsur) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx:84]: You can map the interactions to array of contacts arrays and then reduce it to single array

**Review by NaamanTsur — CHANGES_REQUESTED** [review]

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx [hunk 10]
```diff
@@ -76,7 +79,25 @@ const InteractionEventForm: React.FC<Props> = (
...
+        interaction[InteractionEventDialogFields.CONTACTS].forEach((contact: Contact) =>
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx:85]: Cant you just map the array?

File: client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx (hunk 2 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx:71]: we moved this functionality to useTabManagment in order to call change tab function only after the tab saved completly

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/useInteractionsForm.ts [hunk 11]
```diff
@@ -7,156 +7,97 @@ import Contact from 'models/Contact';
...
+                            user: userId,
+                            investigation: epidemiologyNumber
+                        });
+                        loadInteractions();
+                        closeEditDialog();
                     }
+                }).catch((error) => {
+                        logger.error({
+                            service: Service.CLIENT,
+                            severity: Severity.LOW,
```
Comment (MayaSasson) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/useInteractionsForm.ts:57]: high

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 7 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:59]: what do you mean ?

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 12]
```diff
@@ -1,14 +1,48 @@
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    const checkDuplicateIdsForInteractions = (idsToCheck: IdToCheck[]) => {
+        const trimmedIds = idsToCheck.filter((id: IdToCheck) => id.id !== undefined && id.id !== null && id.id !== '');
+        const duplicateIds = trimmedIds.filter((id, index: number) => 
+            trimmedIds.findIndex((IdToCheck) => IdToCheck.id === id.id && IdToCheck.serialId !== id.serialId) !== -1);
```
Comment (MayaSasson) [inline client/src/Utils/vendor/useDuplicateContactId.ts:35]: same as above

File: client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx [hunk 13]
```diff
@@ -99,6 +99,9 @@ const ClinicalDetails: React.FC<Props> = ({ id }: Props): JSX.Element => {
                     investigation: epidemiologyNumber,
                     user: userId
                 });
+                ClinicalDetailsSchema(validationDate).isValid(values).then(valid => {
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx:102]: why move it to the then block, maybe you should create finally block?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 5 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:28]: the right condition is id
i will change it

File: client/src/Utils/vendor/useDuplicateContactId.ts (hunk 1 above)
Comment (ItayBenMoshe) [reply client/src/Utils/vendor/useDuplicateContactId.ts:22]: We tried to filter twice on the same array. It didn't wotk for us

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 14]
```diff
@@ -1,14 +1,48 @@
...
+    serialId?: number 
+}
+
 const useDuplicateContactId = () => {
     
-    const {alertWarning} = useCustomSwal();
+    const { alertWarning } = useCustomSwal();
+    const epidemiologyNumber = useSelector<StoreStateType, number>(state => state.investigation.epidemiologyNumber);
+    const userId = useSelector<StoreStateType, string>(state => state.user.id);
+
```
Comment (MayaSasson) [inline client/src/Utils/vendor/useDuplicateContactId.ts:20]: The difference between the functions' goals isnt clear - maybe change the first function's name.
In addition, why checking distinct ids only on the second one?

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx [hunk 15]
```diff
@@ -58,34 +49,7 @@ const ContactQuestioning: React.FC<Props> = ({id}: Props): JSX.Element => {
...
-                    workflow: 'Saving all contacts',
-                    step: `got the following error from the server: ${err}`,
-                    user: userId,
-                    investigation: investigationId
-                });
-            });
-        } else {
-            alertWarning('שים לב! נמצאו מספרי זיהוי זהים! אנא בצע את השינויים הנדרשים');
-        }
+        saveContactQuestioning();
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx:52]: add the logs

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 3 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:70]: i will

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx [hunk 16]
```diff
@@ -58,34 +49,7 @@ const ContactQuestioning: React.FC<Props> = ({id}: Props): JSX.Element => {
 
     const saveContacted = (event: React.ChangeEvent<{}>) => {
         event.preventDefault();
-        if(checkAllContactsForDuplicateIds()) {
-            setFormState(investigationId, id, true);
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx:62]: why deleting this line?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 8 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:87]: done

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 7 above)
Comment (MayaSasson) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:59]: export הושלם התחקור to const

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts [hunk 17]
```diff
@@ -20,31 +21,35 @@ import {
...
-            interactedContact.identificationNumber = '';
-            const changedInteractedContact = allContactedInteractions.findIndex(currContact => currContact.id === interactedContact.id);
-            allContactedInteractions.splice(changedInteractedContact, 1, interactedContact);
-            setAllContactedInteractions(allContactedInteractions);
-            handleDuplicateIdsError(interactedContact.identificationNumber, userId, epidemiologyNumber);
+    const saveContact = (interactedContact: InteractedContact) : boolean => {
+        const isDuplicateId = allContactedInteractions.some((contact: InteractedContact) => 
+            contact.id !== interactedContact.id &&
+            interactedContact.identificationNumber !== '' && 
+            interactedContact.identificationNumber !== null &&
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:35]: you can replace these with 
```suggestion
            !interactedContact.identificationNumber  &&
```

Comment (RoeiRom) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:35]: @MayaSasson the genius

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts [hunk 18]
```diff
@@ -20,31 +21,35 @@ import {
...
-            investigation: epidemiologyNumber
-        });
-        if(!checkForSpecificDuplicateIds(interactedContact.identificationNumber, interactedContact.id)) {
-            interactedContact.identificationNumber = '';
-            const changedInteractedContact = allContactedInteractions.findIndex(currContact => currContact.id === interactedContact.id);
-            allContactedInteractions.splice(changedInteractedContact, 1, interactedContact);
-            setAllContactedInteractions(allContactedInteractions);
-            handleDuplicateIdsError(interactedContact.identificationNumber, userId, epidemiologyNumber);
+    const saveContact = (interactedContact: InteractedContact) : boolean => {
+        const isDuplicateId = allContactedInteractions.some((contact: InteractedContact) =>
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:32]: why not using checkDuplicateIds here? you can send [interactedContact, ...allContactedInteractions] to the function

File: client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx [hunk 19]
```diff
@@ -196,6 +195,7 @@ const ExposuresAndFlights : React.FC<Props> = ({ id }: Props): JSX.Element => {
           investigation: investigationId,
           user: userId
       });
+      setFormState(investigationId, id, true);
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx:198]: why moving it to the then?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 7 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:59]: DONE

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/ContactDateCard.tsx [hunk 20]
```diff
@@ -63,6 +64,7 @@ const ContactDateCard: React.FC<Props> = (props: Props) => {
 }
 
 interface Props {
+    allInteractions: Interaction[];
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/ContactDateCard.tsx:67]: I wonder why moving it as prop and not creating context to the 4th and 5th tab?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 21]
```diff
@@ -18,16 +19,19 @@ const fileEndings = [
...
-    contactEvent:number;
+    contactEvent: number;
     onSave: () => void;
+    allInteractions: Interaction[];
 }
-const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
+const ContactUploader = ({contactEvent, onSave, allInteractions} :ExcelUploaderProps) => {
+    const COMPLETE_CONTACT_QUESTIONING_STATUS = 'הושלם התחקור';
+    const existingContacts = allInteractions.map(interaction => interaction.contacts).flat();
+    const existingContactsIds : string[] = existingContacts.map(contact => contact.idNumber).filter(id => id) as string[];
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:29]: is the casting at the end neccessary?

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 22]
```diff
@@ -50,33 +55,44 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
-                        severity: Severity.LOW,
-                        workflow: 'Saving Contacted People Excel',
-                        step: `got error from server: ${error}`,
-                        investigation: epidemiologyNumber,
-                        user: userId
-                    })
-                    alertError('שגיאה בשמירת הנתונים');
+            const excelContactsIds = contacts.map((contact) => contact.identificationNumber)
+                                             .filter(id => id);
+            if(contacts.some((contact) => contact.contactStatus == COMPLETE_CONTACT_QUESTIONING_STATUS)) {
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:60]: ===

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 6 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx]: done

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx [hunk 23]
```diff
@@ -50,33 +55,44 @@ const ContactUploader = ({contactEvent, onSave}:ExcelUploaderProps) => {
...
+                    service: Service.CLIENT,
+                    severity: Severity.LOW,
+                    workflow: 'Saving Contacted People Excel',
+                    step: `trying to save excel with complete status contact`,
+                    investigation: epidemiologyNumber,
+                    user: userId
                 })
+                alertError('לא ניתן לטעון מגע בסטטוס הושלם התחקור, שנו סטטוס זה ונסו שוב');
+            }
+            else if(!checkDuplicateIds(excelContactsIds.concat(existingContactsIds))) {
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:71]: I dont know but maybe in order to create misunderstanding you should use [...arr1, ...arr2] because the concat function can make somone new to think that .concat changes one of the arrays

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/InteractionCard.tsx [hunk 24]
```diff
@@ -116,7 +116,12 @@ const InteractionCard: React.FC<Props> = (props: Props) => {
                         <Grid item xs={12}>
                             <div className={classes.excelControllers}>
                                 <ExcelFormatDownloader />
-                                {interaction.id && <ContactUploader contactEvent={interaction.id} onSave={props.loadInteractions} />}
+                                {interaction.id && 
+                                <ContactUploader 
+                                    contactEvent={interaction.id}
+                                    onSave={props.loadInteractions}
+                                    allInteractions = {allInteractions}
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/InteractionCard.tsx:123]: remove space

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts (hunk 4 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/useInteractionsForm.ts (hunk 11 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/useInteractionsForm.ts]: fixed

File: client/src/Utils/vendor/useDuplicateContactId.ts (hunk 12 above)
Comment (ItayBenMoshe) [reply client/src/Utils/vendor/useDuplicateContactId.ts:39]: fixed

File: client/src/Utils/vendor/useDuplicateContactId.ts (hunk 14 above)
Comment (ItayBenMoshe) [reply client/src/Utils/vendor/useDuplicateContactId.ts:20]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionsTab.tsx [hunk 25]
```diff
@@ -71,6 +71,7 @@ const InteractionsTab: React.FC<Props> = (props: Props): JSX.Element => {
                 {
                     getDatesToInvestigate(doesHaveSymptoms, symptomsStartDate, coronaTestDate).reverse().map(date =>
                         <ContactDateCard
+                            allInteractions = {interactions}
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionsTab.tsx:74]: look at the context comment

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx [hunk 26]
```diff
@@ -295,6 +294,9 @@ const PersonalInfoTab: React.FC<Props> = ({ id }: Props): JSX.Element => {
                             step: `got errors updating status: ${error}`,
                         }))
                 }
+                personalInfoValidationSchema.isValid(data).then(valid => {
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx:297]: why moving it to the then block?

File: client/src/components/App/Content/InvestigationForm/TabManagement/useTabManagement.ts [hunk 27]
```diff
@@ -1,49 +1,71 @@
...
-    }, [currentInvestigationValidations])
+    }, [tabsValidations])
 
     const setNextTab = (nextTabId: number) => {
         //@ts-ignore
         nextTab.current = nextTabId;
     }
 
+    const isInvestigationValid = () => {
+        return !(tabsValidations.slice(0, lastTabDisplayedId + 1).some((tabValidation) => !tabValidation));
```
Comment (RoeiRom) [inline client/src/components/App/Content/InvestigationForm/TabManagement/useTabManagement.ts:61]: why you need to slice the tabsValidations array?

**Review by RoeiRom — CHANGES_REQUESTED** [review]

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx (hunk 9 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx:84]: done

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx (hunk 10 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionEventForm/InteractionEventForm.tsx:85]: done

File: client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx [hunk 28]
```diff
@@ -111,7 +73,8 @@ const InvestigationForm: React.FC = (): JSX.Element => {
                                     type="submit"
                                     form={`form-${currentTab}`}
                                     test-id={currentTab === lastTabDisplayedId ? 'endInvestigation' : 'continueToNextStage'}
-                                    onClick={handleNextPageClick}>
+                                    onClick={() => setNextTab(currentTab + 1)}
```
[the remark this answers is not in the document]
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/InvestigationForm.tsx:76]: moevd to useTabManagement

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx (hunk 26 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx:297]: I will move it to finally block as you suggested above

File: client/src/components/App/Content/InvestigationForm/TabManagement/useTabManagement.ts (hunk 27 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/useTabManagement.ts:61]: we just moved it there as it was

Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/useTabManagement.ts:61]: I guess it because we don't know the length of the array. it depends on areThereContacts variable

File: client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx (hunk 19 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ExposuresAndFlights/ExposuresAndFlights.tsx:198]: setFormState is responsible for moving tabs and i want to move tab after saving the data

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/ContactDateCard.tsx (hunk 20 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/ContactDateCard.tsx:67]: I agree
For now it doesn't matter

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionsTab.tsx (hunk 25 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/InteractionsTab.tsx:74]: maybe after solving the bugs we should change it as well

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts (hunk 17 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:35]: I'll do it as you suggest below

Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts:35]: you can close it

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx (hunk 15 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/ContactQuestioning.tsx:52]: what do you mean?

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 29]
```diff
@@ -1,14 +1,48 @@
...
+            handleDuplicateIdsError(duplicateIds);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    const checkDuplicateIdsForInteractions = (idsToCheck: IdToCheck[]) => {
+        const trimmedIds = idsToCheck.filter((id: IdToCheck) => id.id !== undefined && id.id !== null && id.id !== '');
+        const duplicateIds = trimmedIds.filter((id, index: number) =>
```
[the remark this answers is not in the document]
Comment (ItayBenMoshe) [reply client/src/Utils/vendor/useDuplicateContactId.ts:38]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx (hunk 13 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ClinicalDetails/ClinicalDetails.tsx]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts (hunk 17 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts (hunk 18 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/ContactQuestioning/useContactQuestioning.ts]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/InteractionCard.tsx (hunk 24 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/InteractionCard.tsx]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx (hunk 26 above)
Comment (ItayBenMoshe) [reply client/src/components/App/Content/InvestigationForm/TabManagement/PersonalInfoTab/PersonalInfoTab.tsx]: fixed

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 3 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx]: done

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 30]
```diff
@@ -1,14 +1,48 @@
...
+        } else {
+            return false;
+        }
+    }
+
+    const checkDuplicateIdsForInteractions = (idsToCheck: IdToCheck[]) => {
+        const trimmedIds = idsToCheck.filter((id: IdToCheck) => id.id !== undefined && id.id !== null && id.id !== '');
+        const duplicateIds = trimmedIds.filter((id, index: number) => 
+            trimmedIds.findIndex((IdToCheck) => IdToCheck.id === id.id && IdToCheck.serialId !== id.serialId) !== -1);
+        const distinctDuplicateIds = duplicateIds.filter((id1, index) => duplicateIds.findIndex(id2 => id1.id === id2.id) === index)
```
[the remark this answers is not in the document]
Comment (eladyashar) [reply client/src/Utils/vendor/useDuplicateContactId.ts:40]: done

File: client/src/Utils/vendor/useDuplicateContactId.ts [hunk 31]
```diff
@@ -1,14 +1,48 @@
...
+        if (distinctDuplicateIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (eladyashar) [reply client/src/Utils/vendor/useDuplicateContactId.ts:41]: you are right, i changed it

File: client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx (hunk 23 above)
Comment (eladyashar) [reply client/src/components/App/Content/InvestigationForm/TabManagement/InteractionsTab/ContactDateCard/InteractionCard/ExcelUploader/ContactUploader.tsx:71]: but its not mutating the array, like almost any array functions in java script concat is immutable, i think that your option isn't better, as same i can say that someone new wont be familiar with concat immutability some new might not know what spread does.
if its a common standart i will change it, but i think its more clear and readable the way it is now
